File: backend/models/subscription.py
import enum
import uuid
from datetime import date, datetime

from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    Enum,
    ForeignKey,
    Integer,
    Numeric,
    String,
    Text,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from database.base import Base


class SubscriptionStatus(str, enum.Enum):
    ACTIVE = "ACTIVE"
    EXPIRED = "EXPIRED"
    PAUSED = "PAUSED"
    CANCELLED = "CANCELLED"


class SubscriptionPaymentStatus(str, enum.Enum):
    PAID = "PAID"
    PENDING = "PENDING"
    FAILED = "FAILED"
    REFUNDED = "REFUNDED"


class Subscription(Base):
    """A client's subscription to a plan, tracking its full lifecycle.

    Plan details (name, price, currency, duration) are copied into snapshot
    fields at creation time and are immutable thereafter, so later edits to
    the referenced SubscriptionPlan never alter historical subscriptions.
    Subscriptions are never deleted (Version-1); a client accumulates one row
    per subscription over time. Trainers must never be given plan_price,
    plan_currency, payment_status, or notes (see docs/subscription views).
    """

    __tablename__ = "subscriptions"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    client_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("clients.id"), nullable=False
    )
    subscription_plan_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("subscription_plans.id"), nullable=False
    )

    # Snapshot of the SubscriptionPlan at purchase time; immutable after
    # creation so plan edits never rewrite history.
    plan_name: Mapped[str] = mapped_column(String(100), nullable=False)
    plan_price: Mapped[float] = mapped_column(Numeric(10, 2), nullable=False)
    plan_currency: Mapped[str] = mapped_column(String(3), nullable=False)
    plan_duration_days: Mapped[int] = mapped_column(Integer, nullable=False)
    # Snapshot of SubscriptionPlan.sessions_per_week at purchase time (Task-21);
    # nullable because the source plan field is itself optional.
    plan_sessions_per_week: Mapped[int] = mapped_column(Integer, nullable=True)

    start_date: Mapped[date] = mapped_column(Date, nullable=False)
    end_date: Mapped[date] = mapped_column(Date, nullable=False)

    status: Mapped[SubscriptionStatus] = mapped_column(
        Enum(SubscriptionStatus, name="subscription_status"), nullable=False
    )
    payment_status: Mapped[SubscriptionPaymentStatus] = mapped_column(
        Enum(SubscriptionPaymentStatus, name="subscription_payment_status"),
        nullable=False,
    )

    auto_renew: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    # Annotation kept non-Optional (with nullable=True set explicitly) because
    # SQLAlchemy 2.0.36's Mapped[X | None] resolution crashes on Python 3.14;
    # the column is nullable at the DB/ORM level regardless.
    notes: Mapped[str] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    client: Mapped["Client"] = relationship(  # noqa: F821
        "Client", foreign_keys=[client_id], back_populates="subscriptions"
    )
    subscription_plan: Mapped["SubscriptionPlan"] = relationship(  # noqa: F821
        "SubscriptionPlan",
        foreign_keys=[subscription_plan_id],
        back_populates="subscriptions",
    )

    def __repr__(self) -> str:
        return (
            f"Subscription(id={self.id!r}, client_id={self.client_id!r}, "
            f"status={self.status!r})"
        )

File: frontend/src/types/subscription.ts
import type { PaginatedResponse } from "@/types/dashboard"

export type SubscriptionStatus = "ACTIVE" | "EXPIRED" | "PAUSED" | "CANCELLED"
export type SubscriptionPaymentStatus = "PAID" | "PENDING" | "FAILED" | "REFUNDED"

export interface Subscription {
  id: string
  client_id: string
  subscription_plan_id: string
  plan_name: string
  plan_price: number
  plan_currency: string
  plan_duration_days: number
  plan_sessions_per_week: number | null
  start_date: string
  end_date: string
  status: SubscriptionStatus
  payment_status: SubscriptionPaymentStatus
  auto_renew: boolean
  notes: string | null
  created_at: string
  updated_at: string
}

export type PaginatedSubscriptions = PaginatedResponse<Subscription>

export interface SubscriptionCreateInput {
  client_id: string
  subscription_plan_id: string
  start_date?: string | null
  auto_renew: boolean
  notes?: string | null
}

/** Only status, payment_status, end_date, auto_renew, and notes are editable -
 * client_id, subscription_plan_id, and every plan-snapshot/start_date field
 * are immutable server-side (Task-16.3.1) and are never sent here. */
export interface SubscriptionUpdateInput {
  status?: SubscriptionStatus
  payment_status?: SubscriptionPaymentStatus
  end_date?: string
  auto_renew?: boolean
  notes?: string | null
}

/** Summary shape returned by `GET /subscriptions/my-subscriptions` - narrower
 * than `Subscription` (no ids, duration, or sessions-per-week). */
export interface ClientSubscription {
  id: string
  plan_name: string
  plan_price: number
  plan_currency: string
  payment_status: SubscriptionPaymentStatus
  status: SubscriptionStatus
  start_date: string
  end_date: string
}

/** Deliberately excludes plan_price, plan_currency, payment_status, notes,
 * auto_renew, and subscription ids - trainers must never receive financial or
 * historical subscription data (see backend SubscriptionEligibility). */
export interface SubscriptionEligibility {
  client_id: string
  plan_name: string
  status: SubscriptionStatus
  end_date: string
  can_schedule_sessions: boolean
}

export interface SubscriptionPlan {
  id: string
  name: string
  description: string | null
  duration_days: number
  price: number
  currency: string
  max_sessions_per_month: number | null
  sessions_per_week: number | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface SubscriptionPlanCreateInput {
  name: string
  description?: string | null
  duration_days: number
  price: number
  currency: string
  max_sessions_per_month?: number | null
  sessions_per_week?: number | null
}

/** name, duration_days, and currency are immutable server-side
 * (Task-16.3.1) and are never sent here. */
export interface SubscriptionPlanUpdateInput {
  description?: string | null
  price?: number
  max_sessions_per_month?: number | null
  sessions_per_week?: number | null
  is_active?: boolean
}
